plugins/woocommerce/includes/admin/helper/class-wc-helper-subscriptions-api.php

```diff
@@ -136,9 +164,83 @@ public static function deactivate( $request ) {
 			wp_send_json_error(
 				array(
 					'message' => __( 'There was an error deactivating your subscription. Please try again.', 'woocommerce' )
-				)
+				),
+				400
+			);
+		}
+	}
+
+	/**
+	 * Install a WooCommerce.com produc
+	 */
+	public static function install( $request ) {
+		$product_key = $request->get_param('product_key');
+		$subscriptions = WC_Helper::get_subscription( $product_key );
+
+		if ( empty( $subscriptions ) ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'We couldn\'t find this subscription.', 'woocommerce' )
+				),
+				404
+			);
+		}
+
+		if ( $subscriptions['expired'] === true ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'This subscription has expired.', 'woocommerce' )
+				),
+				402
 			);
 		}
+
+		if ( $subscriptions['maxed'] === true ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'All licenses for this subscription are already in use.', 'woocommerce' )
+				),
+				402
+			);
+		}
+
+		$activation_success = WC_Helper::activate_helper_subscription( $product_key );
+		if ( ! $activation_success ) {
+			wp_send_json_error(
+				array(
+					'message' => __( 'We couldn\'t activate your subscription.', 'woocommerce' )
+				),
+				400
+			);
+		}
+
+		$product_id = $subscriptions['product_id'];
+
+		// Delete any existing state for this product.
+		$state = WC_WCCOM_Site_Installation_State_Storage::get_state( $product_id );
+		if ( $state !== null ) {
+			WC_WCCOM_Site_Installation_State_Storage::delete_state( $state );
+		}
+
+		// Run the installation.
+		$installation_manager = new WC_WCCOM_Site_Installation_Manager( $product_id, $product_id );
```

Comment:
That's a good catch. I can't quite figure out what a good idempotency key would be. The other two occasions actually get this value from the front end. I think it must be something like to prevent situations where a user clicks on the **Update** or **Install** button two times.

I'll think about this and maybe log it as an issue.